ui/components/multichain/edit-accounts-modal/edit-accounts-modal.tsx

```diff
@@ -17,65 +22,180 @@ import {
+import {
+  addMorePermittedAccounts,
+  removePermissionsFor,
+  removePermittedAccount,
+} from '../../../store/actions';
+import { SubjectsType } from '../pages/connections/components/connections.types';
+
+const defaultAllowedAccountTypes = [EthAccountType.Eoa, EthAccountType.Erc4337];
 
 type EditAccountsModalProps = {
   onClose: () => void;
+  onClick: () => void;
   allowedAccountTypes?: KeyringAccountType[];
 };
 
-const defaultAllowedAccountTypes: KeyringAccountType[] = [
-  EthAccountType.Eoa,
-  EthAccountType.Erc4337,
-];
-
-export const EditAccountsModal = ({
+export const EditAccountsModal: React.FC<EditAccountsModalProps> = ({
   onClose,
+  onClick,
   allowedAccountTypes = defaultAllowedAccountTypes,
-}: EditAccountsModalProps) => {
+}) => {
   const t = useI18nContext();
   const accounts = useSelector(getUpdatedAndSortedAccounts);
   const internalAccounts = useSelector(getInternalAccounts);
+  const dispatch = useDispatch();
 
   const mergedAccounts: MergedInternalAccount[] = useMemo(() => {
     return mergeAccounts(accounts, internalAccounts).filter(
       (account: InternalAccount) => allowedAccountTypes.includes(account.type),
     );
-  }, [accounts, internalAccounts, allowedAccountTypes]); // Add allowedAccountTypes to dependency array
+  }, [accounts, internalAccounts, allowedAccountTypes]);
+
+  const activeTabOrigin = useSelector(getOriginOfCurrentTab);
+  const subjects = useSelector(getPermissionSubjects);
+  const connectedAccounts = useSelector((state: any) =>
+    getOrderedConnectedAccountsForConnectedDapp(state, activeTabOrigin).filter(
+      (account: InternalAccount) => isEvmAccountType(account.type),
+    ),
+  );
+
+  const connectedAccountsAddresses = connectedAccounts.map(
+    (account: InternalAccount) => account.address,
+  );
+
+  const [selectedAccounts, setSelectedAccounts] = useState<string[]>(
+    connectedAccountsAddresses,
+  );
+
+  const handleAccountClick = (address: string) => {
+    const index = selectedAccounts.indexOf(address);
+    let newSelectedAccounts: string[];
+
+    if (index === -1) {
+      newSelectedAccounts = [...selectedAccounts, address];
+    } else {
+      newSelectedAccounts = selectedAccounts.filter(
+        (account) => account !== address,
+      );
+    }
+    setSelectedAccounts(newSelectedAccounts);
+  };
+
+  const disconnectAllAccounts = () => {
+    const subject = (subjects as SubjectsType)[activeTabOrigin];
+    if (subject) {
+      const permissionMethodNames = Object.values(subject.permissions).map(
+        ({ parentCapability }: { parentCapability: string }) =>
+          parentCapability,
+      ) as string[];
+      if (permissionMethodNames.length > 0) {
+        const permissionsRecord: Record<string, string[]> = {
+          [activeTabOrigin]: permissionMethodNames,
+        };
+
+        dispatch(
+          removePermissionsFor(
+            permissionsRecord as Record<string, NonEmptyArray<string>>,
+          ),
```

Comment:
I see a lot of type casting in this method, which can be a code smell. Can we infer the types?

Reply:
Updated it

Reply:
unfortunately the `.length > 0` check doesn't help typescript infer the NonEmptyArray. I've removed most of the type casts but had to leave that one

https://github.com/MetaMask/metamask-extension/pull/26635/commits/8984bab4d07d3cf5ceb8ba43d046b608a9b81bc9